### ISSUE TITLE
Save selection for further use

### ISSUE THREAD

**Issue by pdemilly** [tabulapdf/tabula#483]
Great work on tabula and the new UI

I am using tabula to extract disparate ledger info from different system into our system for consolidation. Those systems generate altogether very different format but each one is the same from month to month. I would like to be able to save a selection template and recall it at will. Let me know if that is possible

I would also like to only use the selection viewer. We already have our PDF stored on our backends. We just need to process them and we can use tabula-java for that. But the viewer and its selection will make the creation of the template very easy. Any idea if it is possible? or in the plans

Thanks

**Comment by jeremybmerrill** [comment]
Hi Pascal, both of those features are on the TODO list, but there's no set timeline. (Tabula is an off-hours project for all of us...)

The good news is that a lot of the preliminary work is done, so I think it'd be totally possible for you to modify the Tabula UI to add the features. We'd gladly accept a pull request adding either one and would be happy to advise in the process.

On saving selections, there are two options:
-  you can already select the JSON download format. You'll get back a definition of the areas you've selected. There's not currently a way to re-upload that JSON file, but it'd be fairly easy to add one; the "autodetect" code involves generating a similar JSON file of table areas. You'd just have to write the code to upload the file and then call the autodetect methods with the JSON from the uploaded file.
- alternatively, you can already select the Script download format, which generates a tabula-java (or maybe tabula-extractor, in which case we'll fix that very shortly) script for your selection that could then be applied to any number of input documents.

On the separation of the processing API and the UI, a lot of the preliminary work is already done. [webapp/tabula_web.rb](https://github.com/tabulapdf/tabula/blob/master/webapp/tabula_web.rb) functions as an API layer that the UI interacts with in a mostly-decoupled sort of way. (The sticking point is the async job processing.) We eventually aim to separate the API into https://github.com/tabulapdf/tabula-api, but that project is sort of dormant for now. You could probably rewrite the https://github.com/tabulapdf/tabula/blob/master/webapp/tabula_web.rb#L229-L326 to simply write out to disk the script/JSON templates (instead of the current functionality, which is to use tabula-java to get the table data and return it in the requested format).

In short, what you're looking for isn't implemented yet, but is doable because we've had this goal in mind. If you wanted to work on some part of this, we'd be happy to advise.

**Comment by pdemilly** [comment]
I could give it a try. However I am a java and angular programmer. Don't know jruby or Backbone and from I see they are intimately linked. I would like to make a directive of the viewer though. Any tips on what files I need to look at?

**Comment by jeremybmerrill** [comment]
I don't quite understand what you mean when you say you want to "make a directive of the viewer"?

Basically all the logic is in https://github.com/tabulapdf/tabula/blob/master/webapp/static/js/pdf_view.js and the HTML and templates are in https://github.com/tabulapdf/tabula/blob/master/webapp/index.html. 

The rectangular selectors libraries are in https://github.com/tabulapdf/tabula/blob/master/webapp/static/js/ -- but you'd definitely want to refer to the pdf_view.js file to see how to translate between the locations on the page (from the rectangular selector library) to locations on the PDF. The scaling can be a little touchy...

Let me know if you want more guidance; I'm happy to help.

**Comment by jazzido** [comment]
> I don't quite understand what you mean when you say you want to "make a directive of the viewer"?

I'm pretty sure @pdemilly referes to an Angular [`directive`](https://docs.angularjs.org/guide/directive)

**Comment by jeremybmerrill** [comment]
Aha! I've never done anything more than the first tutorial in Angular.

The most important thing that the viewer does is create a Backbone model object for the Selection when you're done dragging your mouse around to make your selection. That code happens here: https://github.com/tabulapdf/tabula/blob/master/webapp/static/js/pdf_view.js#L733-L753 and the Selection model (which is then what gets sent to tabula-java for data, or transformed into a JSON/shell script representation of your selection) is in the same file here https://github.com/tabulapdf/tabula/blob/master/webapp/static/js/pdf_view.js#L52-L107.

**Comment by pdemilly** [comment]
Yes I meant an angular directive (That's what I know to program) which could become a webcomponent. Looking at the code it looks to me that you are not using a pdf viewer like pdfviewer or pdfjs from mozilla but instead generating images (thumbnails and views) and displaying them. I am not sure how you get the list of images then. Must be a call to the backend, I guess: this.pdf_document.fetch(); and this.pdf_document.page_collection.fetch() which from reading about Backbone seems to make a rest call to backend. So my question: Do you have an API reference of the backend?

**Comment by jeremybmerrill** [comment]
Correct, we bake out the PDF to images and then display those. (Manuel has been thinking about incorporating PDF.js for a long time...)

There isn't really an API, so much as a JSON file created by the uploader process. It's created by this [one job](https://github.com/tabulapdf/tabula/blob/master/lib/tabula_job_executor/jobs/generate_document_data.rb) that writes the `/pdfs/PDF_ID/pages.json` file (where PDF_ID is the big hex string in the Tabula URL, e.g. `localhost:9292/pdfs/3ff8f00fd48ef679548ee55086607ba001a43762/pages.json`).

The `this.pdf_document.page_collection.fetch()` method just gets and loads that JSON file.

That JSON file looks like this, with one object per page:

````
[
    {
        "width": ​612.0,
        "height": ​792.0,
        "number": ​1,
        "rotation": ​0,
        "hasText": true
    }
]
````
The uploader's progress bar is waiting on this file (and the thumbnails) to be generated, so pages.json is guaranteed to be present if the progress bar is finished.

**Comment by pdemilly** [comment]
Actually I think it is quite a smart way to do it

How do you do the post to upload the file?

**Comment by jeremybmerrill** [comment]
It's kind of a mess -- and the biggest barrier to really separating the UI from the API. When you upload the file (just an ordinary HTML file input), it gets written to disk, and a few async jobs get performed on it: writing out [thumbnails/images](https://github.com/tabulapdf/tabula/blob/master/lib/tabula_job_executor/jobs/generate_thumbnails.rb), [writing the page metadata](https://github.com/tabulapdf/tabula/blob/master/lib/tabula_job_executor/jobs/generate_document_data.rb) (pages.json) and trying to [automatically detect tables](https://github.com/tabulapdf/tabula/blob/master/lib/tabula_job_executor/jobs/detect_tables.rb) on the page (tables.json). The page metadata job also records the PDF to `workspace.json`, which lists all the files that Tabula knows about.

While those jobs are taking place, the UI pings an endpoint to get the jobs' progress, and displays that on the progress bar. Once it's done, the UI takes you to the extraction page.

Once it's all done, there's no more interaction with the jobs. You could duplicate the upload process by -- using whatever tools you want -- writing the thumbnails, pages.json and tables.json to the path where Tabula expects them, and recording the PDF in workspace.json.

**Comment by prokopyev** [comment]
The update does say "Selections can be *saved* as templates." I've downloaded 1.2.0 for Mac and can't find this feature, can you please help me locate it?

**Comment by jeremybmerrill** [comment]
@prokopyev You should see a button that says Templates. Click that button, then click "Save Selections as Template" under the resulting menu.

**Comment by jeremybmerrill** [comment]
Closing this issue because this has been done for a while and I'm just bad at keeping up with the tickets. 

Feel free to continue to discuss this issue here.
